File: 56.rb
# https://leetcode.com/problems/merge-intervals/
# @param {Integer[][]} intervals
# @return {Integer[][]}
def merge(intervals)
	res = []
	return intervals if intervals.size < 2

	intervals.sort! { |a,b| a[0] <=> b[0] }
	intervals.each_with_index { |curr, i|
		next if !curr

		if i < intervals.size-1
			n = i+1
			while n < intervals.size && intervals[n][0] <= curr[1]
				curr = intervals[i] = [(curr+intervals[n]).min, (curr+intervals[n]).max]
				intervals[n] = nil
				n += 1
			end
		end
	}
	intervals.compact
end

File: js/56.js
// https://leetcode.com/problems/merge-intervals/
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
	if (!intervals || !intervals.length) {
		return intervals || []
	}
	intervals = intervals.sort((a,b) => a[0] - b[0])
	let i = 0
	let res = []
	while (i<intervals.length) {
		if (i === 0) {
			res.push(intervals[i])
		} else {
			if (res[res.length-1][1] >= intervals[i][0]) {
				res[res.length-1][1] = Math.max(intervals[i][1], res[res.length-1][1])
			} else {
				res.push(intervals[i])
			}
		}
		i++
	}
	return res
};
